Ops/src/OpsWorkerGenetic.cpp

```diff
@@ -0,0 +1,231 @@
+
+        //Check convergence
+        if(checkForConvergence()) break;      //Use complex method, it's the same principle
+
+        mpMessageHandler->stepCompleted(mIterationCounter);
+    }
+
+    if(mpMessageHandler->aborted())
+    {
+        mpMessageHandler->printMessage("Optimization was aborted after "+std::to_string(mIterationCounter)+" iterations.");
+    }
+    else if(mIterationCounter == mnMaxIterations)
+    {
+        mpMessageHandler->printMessage("Optimization failed to converge after "+std::to_string(mIterationCounter)+" iterations");
+    }
+    else
+    {
+        mpMessageHandler->printMessage("Optimization converged in parameter values after "+std::to_string(mIterationCounter)+" iterations.");
+    }
+
+    // Clean up
+    finalize();
+
+    return;
+}
+
+
+//! @brief Set number of optimization parameters
+void WorkerGenetic::setNumberOfParameters(size_t value)
+{
+    Worker::setNumberOfParameters(value);
+
+    mChild1.resize(mNumParameters);
+    mChild2.resize(mNumParameters);
+}
+
+
+//! @brief Set crossover probability parameter
+void WorkerGenetic::setCrossoverProbability(double value)
+{
+    mCrossoverProbability = value;
+}
+
+
+//! @brief Set mutation probability parameter
+void WorkerGenetic::setMutationProbability(double value)
+{
+    mMutationProbability = value;
+}
+
+
+//! @brief Auxiliary function for selecting parents for crossover
+void WorkerGenetic::selectParents() {
+    double bestValue = 1e100;
+    for(size_t i=0; i<mNumPoints; ++i) {
+        double value = mObjectives[i]*opsRand();
+        if(value < bestValue) {
+            mParent1 = i;
+            bestValue = value;
+        }
+    }
+
+    bestValue = 1e100;
+    for(size_t i=0; i<mNumPoints; ++i) {
+        double value = mObjectives[i]*opsRand();
+        if(value < bestValue && i != mParent1) {
+            mParent2 = i;
+            bestValue = value;
+        }
+    }
+}
+
+
+//! @brief Auxiliary function for perform crossover
+void WorkerGenetic::crossOver()
+{
+    for(size_t i=0; i<mNumParameters; ++i)
+    {
+        double x1 = mPoints[mParent1][i];
+        double x2 = mPoints[mParent2][i];
+
+        double doCrossOver = opsRand();
+        if(doCrossOver > mCrossoverProbability)
+        {
+            mChild1[i] = x1;
+            mChild2[i] = x2;
+        }
+        else
+        {
+            double mean = (x1+x2)/2.0;
+            double dev = fabs(x2-x1)/2.0;
+
+            mChild1[i] = gaussian(mean,dev,mParameterMin[i],mParameterMax[i]);
+            mChild2[i] = gaussian(mean,dev,mParameterMin[i],mParameterMax[i]);
+        }
+    }
+    mCandidatePoints[mParent1] = mChild1;
+    mCandidatePoints[mParent2] = mChild2;
+}
+
+
+//! @brief Auxiliary function for mutating children
+void WorkerGenetic::mutate()
+{
+    for(size_t i=0; i<mNumPoints; ++i) {
+        for(size_t j=0; j<mNumParameters; ++j) {
+            double doMutation = opsRand();
+            if(doMutation < mMutationProbability) {
```

Comment:
Here the check is <, but above for crossover it was >, is that intentional?

Reply:
Yes, it is correct. When comparing crossover the conditions in the if statements are reversed, so it actually checks for not performing the  crossover.